### PR TITLE
schema hardening

### PR DESCRIPTION
**Pull request by avalanche123**
### Description of your changes

- upgrade gqlgen
- gqlgen disable struct pointers
- make CompositeResourceSpec.resources required
- add .graphqlrc for vscode
- fix presenting all errors as network

The schema hardening allows returning partial results in face of resolver errors:

Before:

image

After:

image

Note how a connection error, and subsequently returning null from the resolver invalidates the entire object. (`nodes: null`) at the top level, vs nested connection.

I could fix resolvers by hand, but switching gqlgen to not generate pointers to structs makes it impossible to return a `null` for a required field.

I have:

- [x] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR, as appropriate.~

### How has this code been tested

Deployed xGQL locally and verified that it worked.
Upgraded my control plane to this version of xgql and interacted with it via console.

### CODE REVIEW

**Comment by tnthornton** [comment]
>I could fix resolvers by hand, but switching gqlgen to not generate pointers to structs makes it impossible to return a null for a required field.

@avalanche123 given that we're moving to copy the structs around, did you happen to see any differences in memory allocation/utilization for the xgql pod in the control plane you were interacting with?

**Review by tnthornton — APPROVED** [review]

File: schema/composite.gql [hunk 1]
```diff
@@ -68,7 +68,7 @@ type CompositeResourceSpec {
   """
   The resources of which this composite resource is composed.
   """
-  resources: KubernetesResourceConnection @goField(forceResolver: true)
+  resources: KubernetesResourceConnection! @goField(forceResolver: true)
```
Comment (negz) [inline schema/composite.gql:71]: How does this work if an XR actually doesn't have any composed resources? It it just saying it has to have a connection, but the connection may have zero resources?

File: gqlgen.yml [hunk 2]
```diff
@@ -14,6 +14,8 @@ model:
 
 # Use []Thing instead of []*Thing
 omit_slice_element_pointers: true
+struct_fields_always_pointers: false
+resolvers_always_return_pointers: false
```
Comment (negz) [inline gqlgen.yml:18]: Could you add some comments to explain why we do this?

File: gqlgen.yml [hunk 3]
```diff
@@ -40,3 +42,15 @@ models:
...
+      ObjectRefs:
+        type: "[]github.com/crossplane/crossplane-runtime/apis/common/v1.TypedReference"
+  ConfigurationRevisionStatus:
+    extraFields:
+      ObjectRefs:
+        type: "[]github.com/crossplane/crossplane-runtime/apis/common/v1.TypedReference"
+  Event:
+    extraFields:
+      InvolvedObjectRef:
+        type: "k8s.io/api/core/v1.ObjectReference"
```
Comment (negz) [inline gqlgen.yml:56]: Comments explaining what these do would be useful. I think it's allowing the type generator code to better generate the types we need, removing the need for us to define our own Events, Secrets, etc right?

File: internal/graph/present/present.go [hunk 4]
```diff
@@ -131,7 +131,7 @@ func convert(err error) error {
 			Reason: err.Error(),
 			Code:   ErrorRetryable,
 		}
-	case errors.As(err, &dead), errors.As(err, &ref):
+	case errors.Is(err, dead), errors.As(err, &ref):
```
Comment (negz) [inline internal/graph/present/present.go:134]: Ah man, I struggle with the UX of `errors.Is` and `errors.As`. Whoops. 😢

**Review by negz — APPROVED** [review]
This looks great, thank you!

File: gqlgen.yml (hunk 3 above)
Comment (avalanche123) [reply gqlgen.yml:56]: that's exactly right, I'll add comments

File: schema/composite.gql (hunk 1 above)
Comment (avalanche123) [reply schema/composite.gql:71]: that's correct, the nodes in connection are optional (this is true for all connections) and total will be 0

File: gqlgen.yml (hunk 2 above)
Comment (avalanche123) [reply gqlgen.yml]: added comments, please let me know if I missed something.

File: gqlgen.yml (hunk 3 above)
Comment (avalanche123) [reply gqlgen.yml]: same as above, comments are now added

File: internal/graph/present/present.go (hunk 4 above)
Comment (avalanche123) [reply internal/graph/present/present.go:134]: at first I thought this was a novel use of `errors.As`, looked cool as heck. the error presenter code here is pretty clean btw, well done!

**Pull request merged**